### DISCUSSION TITLE
[Help]: `loop: true` not working when images have variable width

### DISCUSSION THREAD

**Discussion by pReya** [davidjerleke/embla-carousel#413] [Help] [answered]
Hey David,

thanks for your work on Embla. I think it's a lovely library, and I've used it successfully before. However in my latest React project I have a problem that I can't be able to solve. I am using `embla-carousel-react` in v7.0.5.

I have built a custom React component around Embla, that accepts Images with Captions via a prop. When I use this component in Isolation (in my component gallery/storybook), everything works great (including the `loop` option).

However if I use the component "in production" inside of my project (which uses Astro as a site generator), I always get this error when I load a page containing the Image slider, and the images never "wrap" around (the space to the left of the first image is always empty)

<img width="848" alt="Screenshot 2023-01-02 at 15 19 13" src="https://user-images.githubusercontent.com/4677417/210244014-aa6ff192-5fb3-409e-acc5-1dbddbc6ade6.png">

I am trying to debug this problem further, but I'm a little lost on how to find the culprit/problem. Something in my setup seems to conflict with Embla, as I cannot recreate this in isolation. Any ideas on how to approach this problem?

**Comment by davidjerleke** [comment, marked as answer]
Hi @pReya,

Thank you for your question. My guess would be that there are no slides when you initialize the Embla script. Try only attaching the `emblaRef` when you have any slides and pass `null` when there aren’t any like so:

`ref={slides.length > 0 ? emblaRef : null}`

Let me know if that’s the culprit or not.

Best,
David

**Reply by pReya** [reply]
Thanks for the quick response. Unfortunately this does not seem to be the problem. As I said: It works fine in isolation. I've also tried to use a static oject (inside of the component – so no props at all) for all the slider content – but it's still throwing the same error.

It seems to be something about my setup/environment that trips up Embla :(

**Reply by davidjerleke** [reply]
@pReya I see. It’s hard for me to guess what could be causing this then. 

> and the images never "wrap" around (the space to the left of the first image is always empty)

This could be a consequence of the script failing before the loop positions are calculated for some reason. It could also be because you’re not providing enough slide content inside the Embla container. In these cases Embla should fall back to `loop: false` though, without crashing.

May I ask what the `<lj>` component in the error message is?

**Reply by pReya** [reply]
@davidjerleke I am sorry for providing very little code and information for you to go on. It is a rather intricate project setup. But I've narrowed it down further now, and I can share the important parts (not yet in an isolated CodeSandbox, but maybe I can provide this later, if I can narrow this down any further). My findings for now:

- The error message in the console seems to be unrelated to the "not wrapping/ignoring loop prop". I've managed to get rid of the console error – yet the wrapping still isn't working.

You can see the problematic behavior here: https://dev.fabcity.hamburg/fabcity-interfacer-ui/main/?story=molecules--image-slider
When this page loads for the first time, the images are not wrapping/looping (tested in latest Firefox and Chrome on macOS). However, when you click and open any other component in the side bar / menu and then come back to the Image Slider, the images will wrap correctly. So this only happens on initial/first load it seems.

The corresponding code for this component can be found here: https://gitlab.fabcity.hamburg/software/fabcity-interfacer-ui/-/blob/main/src/components/ImageSlider.tsx

**Reply by davidjerleke** [reply]
@pReya thanks for the additional details. Do you have placeholders for your images that takes up the space before the images have finished loading?

**Reply by pReya** [reply]
@davidjerleke No, nothing like that. Just plain old image tags (even without implicit height/width values).

**Reply by davidjerleke** [reply]
@pReya you should set placeholders to prevent layout thrashing. Your website will jump every time an image loads which can be really annoying for visitors. 

This is why Embla doesn’t behave as expected the first time: Because the images haven’t loaded by the time Embla initializes so they don’t have any dimensions. Embla can’t calculate any loop positions without any dimensions to pick up. And it doesn’t have any automatic way of detecting changes in slide sizes. 

The browser will get the images from its cache the second time you browse the page with the carousel so they will instantly have their correct dimensions when Embla initializes. That’s why it’s working the second time you visit the page with the carousel.

**Reply by pReya** [reply]
@davidjerleke Thanks for clearing that up! So just to make sure I understand this correctly: Embla's loop option will only work, if the actual `<img>` tag has an explicitly set width or height via CSS or width/height attribute? It will not work, if the max height/width of a slide is defined on the parent container?

**Reply by pReya** [reply]
I've now given every `<img>` and explicit height px value. Still, the wrapping behavior hasn't changed :(

**Reply by davidjerleke** [reply]
@pReya Isn’t your carousel horizontal? If that’s the case, you need to give them a width. The width doesn’t have to be a fixed pixel value, it can be percentage too. In any case, the width needs to be set for horizontal scrolling carousels.

**Reply by pReya** [reply]
Thanks a lot, that seems to do the trick, indeed! Not sure I understand the difference between my code, and your documentation regarding [variable sizes](https://www.embla-carousel.com/guides/slide-sizes/#variable-sizes), though. Will try to dig deeper into this topic and see how I can build a heigh-restricted slider with variable image widths.

Thanks for pointing me in the right direction!

**Comment by pReya** [comment]
@davidjerleke Sorry for digging this up again. I am really confused: Is it possible to use Embla (in general) and also the loop option with **dynamic width slides** (e.g. not just variable as in "every slide has a different, but known/static width that we can put into css" but rather dynamic as in "we don't know how big these images are beforehand, b/c they come from user-editable content or an external API")?

If not, then what is the docs section for, where you mention using the `flex: 0 0 auto;` (https://www.embla-carousel.com/guides/slide-sizes/#variable-sizes)? Because currently I am not sure how to use this. Just adding this to the stylesheet will break Emblas functionality (I've created an example here: https://codesandbox.io/s/embla-carousel-variable-widths-loop-problem-es4tz6)

If yes, I really think this should be mentioned in the docs or there should be an example for "dynamic width slides".

**Reply by davidjerleke** [reply]
@pReya I agree that the `flex: 0 0 auto;` part in the docs can be improved.

> If not, then what is the docs section for, where you mention using the `flex: 0 0 auto;`

This will let the content inside the slides decide the size of each slide. In your CodeSandbox, you use images without image aspect ratio placeholders that take up the space before the image has finished loading. So when each image loads it will alter the slide widths from 0 to whatever width the image has. 

Embla doesn’t make any promise about observing your slide sizes for changes. But you can do this manually by calling `embla.reInit()` when your images have loaded and it will work as expected.

Trust me when I say that in most cases, websites have image aspect ratio placeholders to prevent content from jumping and get bad lighthouse scores because of CLS - Content Layout Shifts. Embla will work with different image sizes too but you have to set the aspect ratio placeholders beforehand.

So to summarize, for variable slide sizes to work:

- Your slide content should be in place before Embla is initialized. If your slides contain images, they should have aspect ratio placeholders before the images have loaded.
- Alternatively, manually call `embla.reInit()` when your images have loaded, or your slide sizes have changed as a consequence of you modifying content inside the slides.

Feel free to help out explaining this in an easily digestible manner and we can update the docs. Pull requests are welcome.

Best,
David

**Reply by pReya** [reply]
@davidjerleke Thank you very much for the detailed explanation.

I have tried manually calling `reInit()` after all images have been loaded. Calling it will make the slider work again and correctly center the first image (you can see the layout change) – however it will NOT fix the loop/wraparound display problem. There will still be no wraparound/loop layout even after `reInit()` has been called. Only after you resize the window it will show the wraparound window. So it seems that `reInit()` is not reseting/initializing everything? Is this a possible bug?

Here is an example, where `reInit()` is called after 3 seconds: https://codesandbox.io/s/embla-carousel-variable-widths-loop-problem-es4tz6?file=/src/js/EmblaCarousel.tsx

**Reply by davidjerleke** [reply]
Hi again @pReya,

`reInit` is doing what it’s supposed to so it’s not a bug. 

You can’t set timeout and “guess” when all images have loaded. What if the browser takes more than 3 seconds to load all images? This can happen if the images are large or if you have a slow internet connection?

You have to add load event listeners on every image and fire `reInit` when **all images** have loaded.

Alternatively, you can observe all slide elements with a resize observer and call `reInit` whenever the observer fires.

**Reply by pReya** [reply]
@davidjerleke Yes, I'm aware that this is not production-ready code. Obviously it's not a good idea to wait for a specified amount of time. I only used it trying to make a point: Even if `reInit` is called after all images have loaded (however long it may take), it will not fix/rerender/modify the missing wraparound/loop layout.

**Reply by davidjerleke** [reply]
@pReya I see. I’ll take a look at your timeout sandbox and try to figure out what’s going on then. Can’t promise when though.

**Reply by davidjerleke** [reply]
Hi @pReya,

Thank you for being persevering with this. So I had a couple of minutes to dig into this and managed to figure out why this is happening. I can confirm that this is a bug and it’s related to this:

> Enables infinite looping. Automatically falls back to false if slide content isn't enough to loop.

The explanation of the loop option in the [docs](https://www.embla-carousel.com/api/options/#loop) 👆🏻.

This is what’s happening:

1. You initialize Embla with images that are loading so the slide widths are 0.
2. Embla makes the decision that your carousel doesn’t have enough slide content for the loop effect to work. So it automatically falls back to `loop: false`.
3. After the timeout the images have loaded and the slides have their final widths.
4. You run `reInit()` and Embla actually calculates everything correctly - even the loop positions. So there’s nothing wrong with the `flex: 0 0 auto;` calculations. But since falling back to `loop: false`, Embla mutated the original options with `loop: true` you passed to the `useEmblaCarousel()` hook. 
5. So from now on every `reInit()` you call without the options parameter will be with `loop: false`.
6. What Embla should do is that it should reset the original options to `loop: true` after falling back to `loop: false` when slide content isn’t enough for the loop effect to work. So the next time you run `reInit()` Embla should try `loop: true` again.

I know what needs to be done in the code to fix this so I’ll let you know when I have released a patch with a fix for this. But can’t promise when though.

Cheers,
David

**Reply by pReya** [reply]
@davidjerleke Thanks for following up and confirming that this is indeed a bug. I'll be looking out for a new release ;)

**Reply by davidjerleke** [reply]
Hi @pReya,

I just released [v7.0.6](https://github.com/davidjerleke/embla-carousel/releases/tag/v7.0.6) which includes a **fix** for this bug:

- #414

Try it out and see if it behaves as expected now if you want. Thank you for your contribution 👍.

Cheers,
David
